remote-wallet/src/remote_wallet.rs

```diff
@@ -250,65 +254,234 @@ pub struct RemoteWalletInfo {
+        match self {
+            Self::Unknown => VENDOR_UNKNOWN,
+            Self::Ledger => VENDOR_LEDGER,
         }
+    }
+}
 
-        let mut derivation_path = DerivationPath::default();
-        let mut query_pairs = wallet_path.query_pairs();
-        if query_pairs.count() > 0 {
-            for _ in 0..query_pairs.count() {
-                if let Some(mut pair) = query_pairs.next() {
-                    if pair.0 == "key" {
-                        let key_path = pair.1.to_mut();
-                        if key_path.ends_with('/') {
-                            key_path.pop();
-                        }
-                        derivation_path = DerivationPath::from_key_str(key_path)?;
+impl std::fmt::Display for Vendor {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        let s: &str = self.as_ref();
+        write!(f, "{}", s)
+    }
+}
+
+#[derive(Clone, Debug, Error, PartialEq)]
+pub enum LocaterError {
+    #[error(transparent)]
+    VendorError(#[from] VendorError),
+    #[error(transparent)]
+    PubkeyError(#[from] ParsePubkeyError),
+    #[error(transparent)]
+    DerivationPathError(#[from] DerivationPathError),
+    #[error(transparent)]
+    UriReferenceError(#[from] URIReferenceError),
+    #[error("unimplemented scheme")]
+    UnimplementedScheme,
+    #[error("infallible")]
+    Infallible,
+}
+
+impl From<Infallible> for LocaterError {
+    fn from(_: Infallible) -> Self {
+        Self::Infallible
+    }
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Locater {
+    vendor: Vendor,
+    pubkey: Option<Pubkey>,
+    derivation_path: Option<DerivationPath>,
+}
+
+impl std::fmt::Display for Locater {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        let maybe_path = self.pubkey.map(|p| p.to_string());
+        let path = maybe_path.as_deref().unwrap_or("/");
+        let maybe_query = self.derivation_path.as_ref().map(|d| d.get_query());
+        let maybe_query2 = maybe_query.as_ref().map(|q| &q[1..]);
+
+        let mut builder = URIReferenceBuilder::new();
+        builder
+            .try_scheme(Some("usb"))
+            .unwrap()
+            .try_authority(Some(self.vendor.as_ref()))
+            .unwrap()
+            .try_path(path)
+            .unwrap()
+            .try_query(maybe_query2)
+            .unwrap();
+
+        let uri = builder.build().unwrap();
+        write!(f, "{}", uri)
+    }
+}
+
+impl Locater {
+    pub fn new_from_path<P: AsRef<str>>(path: P) -> Result<Self, LocaterError> {
+        let path = path.as_ref();
+        let uri = URIReference::try_from(path)?;
+        Self::new_from_uri(&uri)
+    }
+
+    pub fn new_from_uri(uri: &URIReference<'_>) -> Result<Self, LocaterError> {
+        let scheme = uri.scheme().map(|s| s.as_str().to_ascii_lowercase());
+        let host = uri.host().map(|h| h.to_string());
+        match (scheme, host) {
+            (Some(scheme), Some(host)) if scheme == "usb" => {
```

Comment:
Should the guts of this case move somewhere more general, to be used for `file` and `ask` in the future?

Reply:
Potentially.  We mostly have it [here](https://github.com/solana-labs/solana/blob/c2becbc0a886c3707a911458904941985ec1cd7b/clap-utils/src/keypair.rs#L143-L169), which was the motivator for the `new_from_uri()` constructor. I was a little torn on where the common thing should live and what it should look like, so just left the API intact for this changeset